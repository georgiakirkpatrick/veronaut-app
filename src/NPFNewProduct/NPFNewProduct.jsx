import React from 'react'
import '@fortawesome/react-fontawesome'
import FormDropdown from '../FormDropdown/FormDropdown'
import FormPage from '../FormPage/FormPage'
import FormPromptWithSub from '../FormPromptWithSub/FormPromptWithSub'
import FormTextInput from '../FormTextInput/FormTextInput'
import FormUrlInput from '../FormUrlInput/FormUrlInput'
import NPFFooter from '../NPFFooter/NPFFooter'
import currencies from '../CURRENCIES'
import formData from '../FORM_DATA'

const NPFNewProduct = ({
  brandId = 1,
  brandArray = [],
  currentPage = 1,
  newProductFields = {
    name: '', 
    url: '' ,
    categoryId: 1,
    featureImageUrl: '',
    currency: '',
    price: '',
    wash: '',
    dry: ''
  },
  setNewProductFields = () => {},
  setPage = () => {}
}) => {
  const findBrandDetails = brand => {
    return brand['id'] === Number(brandId)
  }

  const brandDetailArray = brandArray.filter(findBrandDetails)
  const brandDetails = brandDetailArray.length > 0 ? brandDetailArray[0] : { id: 0, text: '', value: 0 }

  const changeField = event => {
    setNewProductFields({
      ...newProductFields, 
      [event.target.name]: event.target.value
    })
  }

  const currencyId = brandDetails.currencyId ? Number(brandDetails.currencyId) : 0
  const currencyDetails = currencies[currencyId - 1]

  const nextButton = event => {
    event.preventDefault()
    const missingFields = []

    const requiredFields = {
      name: newProductFields.name,
      url: newProductFields.url,
      categoryId: newProductFields.categoryId,
      featureImageUrl: newProductFields.featureImageUrl,
      price: newProductFields.price,
      washId: newProductFields.washId,
      dryId: newProductFields.dryId
    }
      
    Object.keys(requiredFields).forEach(key => {
      if (requiredFields[key] === '' || Number(requiredFields[key]) === 0) {
          missingFields.push(key.replace( /([A-Z])/g, " $1" ).toLowerCase())
      }
    })
        
    if (missingFields.length === 1) {
      alert(`Please complete the '${missingFields[0]}' field`)
    } else if (missingFields.length > 1) {
      alert(`Please complete the following fields: ${missingFields.map(field => `
          ${field}`)}
      `)
    } else if (missingFields.length === 0) {
      setPage(currentPage + 1)
    }
  }

  return (
    <div className='NPFNewProduct'>
      <FormPage title='New Product'>
        <FormPromptWithSub
          prompt={`Brand: ${brandDetails.text}`}
          promptSubtitle=''
        />

        <FormTextInput
          id='product-name'
          name='name'
          prompt='Product name'
          currentValue={newProductFields.name}
          handleChange={event => changeField(event)}
        />

        <FormUrlInput 
          id='product-url'
          name='url'
          prompt='Product URL'
          currentValue={newProductFields.url}
          handleChange={event => changeField(event)}
        />

        <FormDropdown
          id='product-category'
          name='categoryId'
          prompt='Category'
          handleChange={event => changeField(event)}
          currentValue={newProductFields.categoryId}
          options={formData.productCategories}
        />

        <FormUrlInput
          id='feature-image-url'
          name='featureImageUrl'
          prompt='Feature image URL'
          currentValue={newProductFields.featureImageUrl}
          handleChange={event => changeField(event)}
        />

        <FormTextInput
          id='price'
          name='price'
          prompt={`Product price in ${currencyDetails ? currencyDetails.name_plural : ''} (${currencyDetails ? currencyDetails.symbol_native : ''})`}
          currentValue={newProductFields.price}
          handleChange={event => changeField(event)}
        />

        <FormDropdown
          id='washId'
          name='washId'
          prompt='Washing instructions'
          handleChange={event => changeField(event)}
          currentValue={newProductFields.washId}
          options={formData.washOptions}
        />

        <FormDropdown
          id='dryId'
          name='dryId'
          prompt='Drying instructions'
          handleChange={event => changeField(event)}
          currentValue={newProductFields.dryId}
          options={formData.dryOptions}
        />
      </FormPage>

      <NPFFooter 
        buttons='prevNext' 
        previousButton={() => setPage(currentPage - 1)} 
        nextButton={event => {nextButton(event)}}
      /> 
    </div>
  )    
}

export default NPFNewProduct