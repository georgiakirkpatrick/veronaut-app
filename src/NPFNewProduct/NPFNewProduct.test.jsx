import React from 'react'
import { render, screen } from '@testing-library/react'
import { describe, it, expect } from 'vitest'
import NPFNewProduct from './NPFNewProduct'
import currencies from '../CURRENCIES'

describe('NPFNewProduct', () => {
    it ('renders without crashing', () => {
        render(<NPFNewProduct />)
        screen.debug()
    })

    it ('render the UI as expected', () => {
        const tree = render(<NPFNewProduct />)
        expect(tree).toMatchSnapshot()
    })
})