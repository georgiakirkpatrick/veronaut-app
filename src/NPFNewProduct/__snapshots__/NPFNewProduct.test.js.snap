// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NPFNewProduct render the UI as expected 1`] = `
<div
  className="NPFNewProduct"
>
  <section
    className="FormPage"
  >
    <h1>
      New Product
    </h1>
    <form
      autoComplete="off"
      id="form-page"
    >
      <div
        className="FormPromptWithSub"
      >
        <p
          className="FormPromptWithSub__prompt"
        >
          Brand: 
        </p>
        <p
          className="FormPromptWithSub__subtitle"
        >
          
        </p>
      </div>
      <div
        className="FormTextInput"
      >
        <label
          htmlFor="product-name"
        >
          Product name
        </label>
        <input
          id="product-name"
          name="name"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="FormUrlInput"
      >
        <label
          htmlFor="product-url"
        >
          Product URL
        </label>
        <input
          id="product-url"
          name="url"
          onChange={[Function]}
          required={true}
          type="url"
          value=""
        />
      </div>
      <div
        className="FormDropdown"
      >
        <label
          htmlFor="product-category"
        >
          Category
        </label>
        
        <select
          name="categoryId"
          onChange={[Function]}
          required={true}
          value={1}
        >
          <option
            id={0}
            value={0}
          >
            Select a category
          </option>
          <option
            id={1}
            value={1}
          >
            Activewear
          </option>
          <option
            id={2}
            value={2}
          >
            Blazers
          </option>
          <option
            id={3}
            value={3}
          >
            Coats and Jackets
          </option>
          <option
            id={4}
            value={4}
          >
            Dresses
          </option>
          <option
            id={5}
            value={5}
          >
            Facemasks
          </option>
          <option
            id={6}
            value={6}
          >
            Hair Accessories and Hats
          </option>
          <option
            id={7}
            value={7}
          >
            Jeans
          </option>
          <option
            id={8}
            value={8}
          >
            Jumpsuits
          </option>
          <option
            id={9}
            value={9}
          >
            Lingerie
          </option>
          <option
            id={10}
            value={10}
          >
            Pants
          </option>
          <option
            id={11}
            value={11}
          >
            Shoes
          </option>
          <option
            id={12}
            value={12}
          >
            Skirts
          </option>
          <option
            id={13}
            value={13}
          >
            Socks and Tights
          </option>
          <option
            id={14}
            value={14}
          >
            Sweaters
          </option>
          <option
            id={15}
            value={15}
          >
            Swimwear
          </option>
          <option
            id={16}
            value={16}
          >
            Tops
          </option>
        </select>
      </div>
      <div
        className="FormUrlInput"
      >
        <label
          htmlFor="feature-image-url"
        >
          Feature image URL
        </label>
        <input
          id="feature-image-url"
          name="featureImageUrl"
          onChange={[Function]}
          required={true}
          type="url"
          value=""
        />
      </div>
      <div
        className="FormTextInput"
      >
        <label
          htmlFor="price"
        >
          Product price in  ()
        </label>
        <input
          id="price"
          name="price"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="FormDropdown"
      >
        <label
          htmlFor="washId"
        >
          Washing instructions
        </label>
        
        <select
          name="washId"
          onChange={[Function]}
          required={true}
          value=""
        >
          <option
            id={0}
            value={0}
          >
            Select washing instructions
          </option>
          <option
            id={1}
            value={1}
          >
            Not disclosed
          </option>
          <option
            id={2}
            value={2}
          >
            Do not wash
          </option>
          <option
            id={3}
            value={3}
          >
            Machine wash cold (≤85°F/≤30°C)
          </option>
          <option
            id={4}
            value={4}
          >
            Machine wash warm (≤105°F/≤40°C)
          </option>
          <option
            id={5}
            value={5}
          >
            Machine wash hot (≥140°F/≥60°C)
          </option>
          <option
            id={6}
            value={6}
          >
            Hand wash
          </option>
          <option
            id={7}
            value={7}
          >
            Dry clean only
          </option>
          <option
            id={8}
            value={8}
          >
            Do not dry clean
          </option>
          <option
            id={9}
            value={9}
          >
            Do not wash
          </option>
          <option
            id={10}
            value={10}
          >
            Machine wash normal cycle
          </option>
          <option
            id={11}
            value={11}
          >
            Machine wash permanent press
          </option>
          <option
            id={12}
            value={12}
          >
            Machine wash delicate
          </option>
        </select>
      </div>
      <div
        className="FormDropdown"
      >
        <label
          htmlFor="dryId"
        >
          Drying instructions
        </label>
        
        <select
          name="dryId"
          onChange={[Function]}
          required={true}
          value=""
        >
          <option
            id={0}
            value={0}
          >
            Select drying instructions
          </option>
          <option
            id={1}
            value={1}
          >
            Not disclosed
          </option>
          <option
            id={2}
            value={2}
          >
            Hang dry
          </option>
          <option
            id={3}
            value={3}
          >
            Hang dry in the shade
          </option>
          <option
            id={4}
            value={4}
          >
            Dry flat
          </option>
          <option
            id={5}
            value={5}
          >
            Tumble dry
          </option>
          <option
            id={6}
            value={6}
          >
            Tumble dry low heat
          </option>
          <option
            id={7}
            value={7}
          >
            Tumble dry medium heat
          </option>
          <option
            id={8}
            value={8}
          >
            Tumble dry high heat
          </option>
          <option
            id={9}
            value={9}
          >
            Tumble dry normal cycle
          </option>
          <option
            id={10}
            value={10}
          >
            Tumble dry permanent press
          </option>
          <option
            id={11}
            value={11}
          >
            Tumble dry delicate
          </option>
        </select>
      </div>
    </form>
    <div>
      
    </div>
  </section>
  <footer
    className="NPFFooter"
  >
    <button
      className="NPFFooter__previous"
      onClick={[Function]}
      type="button"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-left fa-w-8 NPFFooter__left-arrow"
        data-icon="angle-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 256 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      <span>
        PREVIOUS
      </span>
    </button>
    <button
      className="NPFFooter__next"
      onClick={[Function]}
      type="button"
    >
      <span>
        NEXT
      </span>
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-right fa-w-8 NPFFooter__right-arrow"
        data-icon="angle-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 256 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </button>
  </footer>
</div>
`;
